### ISSUE TITLE
2.3-beta.3 upgrade to 2.3.1 (Debian9), startup issues related to LWM2M

### ISSUE THREAD

**Issue by jkwong38** [emqx/emqx#1426]
#### Environment

- OS: Debian GNU/Linux 9.3 (stretch)
- Erlang/OTP: R20/9.0
- EMQ: 2.3.1

#### Description

I had been running EMQ 2.3-beta.3 successfully for some time and decided to upgrade to the current 2.3.1.

I downloaded the current Debian 9 .deb file and tried to install it by using dpkg -i.
The install completed with only the following warning:

dpkg: warning: unable to delete old directory '/etc/emqttd/lwm2m_xml': Directory not empty

After installation and rebooting the system, EMQ was not able to start, via service or console.

root@server:~# emqttd_ctl status
Node 'emq@127.0.0.1' not responding to pings.

and 

```
root@server:~# service emqttd status
● emqttd.service - LSB: Erlang MQTT Broker
   Loaded: loaded (/etc/init.d/emqttd; generated; vendor preset: enabled)
   Active: failed (Result: exit-code) since Sun 2017-12-31 09:18:10 EST; 39s ago
     Docs: man:systemd-sysv-generator(8)
  Process: 360 ExecStart=/etc/init.d/emqttd start (code=exited, status=1/FAILURE
    Tasks: 1 (limit: 9830)
   CGroup: /system.slice/emqttd.service
           └─637 /usr/lib/emqttd/erts-9.0/bin/epmd -daemon
```

and

```
root@server:~# service emqttd start
Job for emqttd.service failed because the control process exited with error code.
See "systemctl status emqttd.service" and "journalctl -xe" for details.
root@server:~# service emqttd status
● emqttd.service - LSB: Erlang MQTT Broker
   Loaded: loaded (/etc/init.d/emqttd; generated; vendor preset: enabled)
   Active: failed (Result: exit-code) since Sun 2017-12-31 09:19:35 EST; 5s ago
     Docs: man:systemd-sysv-generator(8)
  Process: 1716 ExecStart=/etc/init.d/emqttd start (code=exited, status=1/FAILUR
    Tasks: 1 (limit: 9830)
   CGroup: /system.slice/emqttd.service
           └─637 /usr/lib/emqttd/erts-9.0/bin/epmd -daemon

Dec 31 09:19:13 server systemd[1]: Starting LSB: Erlang MQTT Broker...
Dec 31 09:19:14 server su[1790]: Successful su for emqtt by root
Dec 31 09:19:14 server su[1790]: + ??? root:emqtt
Dec 31 09:19:14 server su[1790]: pam_unix(su:session): session opened for user em
Dec 31 09:19:35 server systemd[1]: emqttd.service: Control process exited, code=e
Dec 31 09:19:35 server systemd[1]: Failed to start LSB: Erlang MQTT Broker.
Dec 31 09:19:35 server systemd[1]: emqttd.service: Unit entered failed state.
Dec 31 09:19:35 server systemd[1]: emqttd.service: Failed with result 'exit-code'
```

the error temporary logs indicate some issue with LWM2M.

```
=====
===== LOGGING STARTED Sun Dec 31 09:01:51 EST 2017
=====
09:01:56.197 [error] You've tried to set lwm2m.port, but there is no setting with that name.
09:01:56.197 [error]   Did you mean one of these?
09:01:56.228 [error]     coap.port
09:01:56.228 [error]     log.dir
09:01:56.228 [error]     mqtt.sn.port
09:01:56.228 [error] You've tried to set lwm2m.certfile, but there is no setting with that name.
09:01:56.228 [error]   Did you mean one of these?
09:01:56.273 [error]     coap.certfile
09:01:56.273 [error]     coap.keyfile
09:01:56.273 [error]     log.error.file
09:01:56.273 [error] You've tried to set lwm2m.keyfile, but there is no setting with that name.
09:01:56.273 [error]   Did you mean one of these?
09:01:56.321 [error]     coap.keyfile
09:01:56.321 [error]     coap.certfile
09:01:56.321 [error]     log.info.file
09:01:56.321 [error] You've tried to set lwm2m.xml_dir, but there is no setting with that name.
09:01:56.321 [error]   Did you mean one of these?
09:01:56.401 [error]     log.dir
09:01:56.401 [error]     mqtt.acl_file
09:01:56.401 [error]     log.console
09:01:56.422 [error] Error generating configuration in phase transform_datatypes
09:01:56.422 [error] Conf file attempted to set unknown variable: lwm2m.xml_dir
09:01:56.422 [error] Conf file attempted to set unknown variable: lwm2m.keyfile
09:01:56.422 [error] Conf file attempted to set unknown variable: lwm2m.certfile
09:01:56.422 [error] Conf file attempted to set unknown variable: lwm2m.port
```

Downgrading back to 2.3-beta3 makes everything work again. 

Sounds like a config issue with some new settings, but I'm unsure how to work with this.

Any help available, please?

**Comment by jkwong38** [comment]
Update:  I was able to get the service started by commenting out the four configuration lines in /etc/emqttd/plugins/emq_lwm2m.conf

I also tried re-installing with the lwm2m_xml deleted, in hopes it would re-install clean again.  But it did not reappear.  Is this directory auto generated?

**Comment by jkwong38** [comment]
Update:  As in my prior update, the service started, but doesn't work properly; so I reverted back to the beta for now.

**Comment by emqplus** [comment]
@jkwong38 The LwM2M project has been moved to EMQ X project, and will be prod-ready in R3.0.

**Comment by grutabow** [comment]
A quick fix in this case is 
`rm /etc/emqttd/plugins/emq_lwm2m.conf`

The root cause is that 

- emq-lwm2m has been included in emqtt v2.3.0-beta3.
- emq-lwm2m has been removed since emqtt v2.3.1. 
- In your upgrade operation from v2.3.0-beta3 to v2.3.1, "/etc/emqttd/plugins/emq_lwm2m.conf" is not deleted as expected which makes v2.3.1 regards that emq-lwm2m is alive either. But no corresponding emq-lwm2m.schema is found which throw exceptions you have seen.

In a word, upgrade from v2.3.0-beta3 requires a clean installation. That is, remove v2.3.0-beta3 thoroughly from disk first.
